### PR TITLE
Improved frontend documents handling

### PR DESCRIPTION
**Pull request by akosyakov**
This PR:
- removes intermediate workspace documents and use Monaco models directly
- filter out fs events in the file resource based on modification timestamp

### CODE REVIEW

File: packages/core/src/common/reference.ts [hunk 1]
```diff
@@ -6,27 +6,63 @@
...
 export interface Reference<T> extends Disposable {
     readonly object: T
 }
 
-export class ReferenceCollection<K, V extends Disposable> {
+export class ReferenceCollection<K, V extends Disposable> implements Disposable {
 
-    protected readonly values = new Map<string, MaybePromise<V>>();
-    protected readonly keyMap = new Map<string, K>();
+    protected readonly _keys = new Map<string, K>();
```
Comment (svenefftinge) [inline packages/core/src/common/reference.ts]: why the `_` prefix?

**Comment by akosyakov** [comment]
With this PR I've added `FileSystem.updateContent` which applies incremental changes if a file is not out of sync otherwise fallbacks to the whole content syncing. With all changes, you should get the better experience of editing large files and saving should happen faster in most of the cases.

The last commit disables dirty diffs and merge conflicts. They are tackled by other PRs. I will remove this commit from the final PR.

Please review.

@lmcbout Could you try to edit your 3mb file with this PR?

File: packages/core/src/common/resource.ts [hunk 2]
```diff
@@ -14,8 +15,8 @@ import { MaybePromise } from "./types";
 
 export interface Resource extends Disposable {
     readonly uri: URI;
-    readContents(options?: { encoding?: string }): Promise<string>;
-    saveContents?(content: string, options?: { encoding?: string }): Promise<void>;
+    readContents(): Promise<string>;
```
Comment (svenefftinge) [inline packages/core/src/common/resource.ts]: no more encoding here?

File: packages/core/src/common/resource.ts [hunk 3]
```diff
@@ -14,8 +15,8 @@ import { MaybePromise } from "./types";
...
+    saveContents?(content: string, changes?: TextDocumentContentChangeEvent[]): Promise<void>;
```
Comment (svenefftinge) [inline packages/core/src/common/resource.ts]: This signature looks strange to me. I'd expect to either set the content or apply changes.

File: packages/core/src/common/reference.ts (hunk 1 above)
Comment (akosyakov) [reply packages/core/src/common/reference.ts]: there is already a property `keys`

**Comment by akosyakov** [comment]
@svenefftinge i've pushed again, removed a content caching in a file resource and sharing a file resource between clients

**Comment by lmcbout** [comment]
Hi @akosyakov , I just tested this PR with a 3M file.
I noticed the file is only sent once when initial loading is done :  GREAT
Auto-Save: off
   When entering new data and save it when I want=> only the range is sent, so less char on the web socket, -> no freezing -> better performance
Auto-Save: on
  Better performance right away, the whole file is not send over the web socket, works like a charm.

One question, may be not related to this PR but when I initially load the file from the b-e to the f-e, I notice on the f-e filesystem the size of the file is 1.5  the size of the file on disk, may be we may have something to do in this area, I don't know why the message is that size. Look at the highlighted file in the picture (4500182 ) and the size in the text area below (3000001).
The header file size should not be half the size of the message payload.

file3m_orig

**Pull request merged**